open_spiel/python/jax/opponent_shaping.py

```diff
@@ -0,0 +1,984 @@
+        correction term by Foerster et al. will be applied.
+        Args:
+            train_state: the agent's train state.
+            batch: a transition batch
+
+        Returns:
+            A tuple (new_train_state, metrics)
+        """
+    loss, policy_grads = jax.value_and_grad(policy_loss)(
+      train_state.policy_params[agent_id], agent_id, batch)
+    correction = lola_correction(train_state, batch)
+    policy_grads = jax.tree_util.tree_map(
+      lambda grad, corr: grad - lola_weight * corr, policy_grads, correction)
+    updates, opt_state = optimizer(policy_grads,
+                                   train_state.policy_opt_states[agent_id])
+    policy_params = optax.apply_updates(train_state.policy_params[agent_id],
+                                        updates)
+    train_state = TrainState(
+      policy_params={
+        **train_state.policy_params, agent_id: policy_params
+      },
+      policy_opt_states={
+        **train_state.policy_opt_states, agent_id: opt_state
+      },
+      critic_params=deepcopy(train_state.critic_params),
+      critic_opt_states=deepcopy(train_state.critic_opt_states),
+    )
+    return train_state, {'loss': loss}
+
+  return update
+
+
+def get_opponent_update_fn(
+  agent_id: int,
+  policy_network: hk.Transformed,
+  optimizer: optax.TransformUpdateFn,
+  num_minibatches: int = 1,
+) -> UpdateFn:
+
+  def loss_fn(params, batch: TransitionBatch):
+
+    def loss(p, states, actions):
+      log_prob = policy_network.apply(p, states).log_prob(actions)
+      return log_prob
+
+    log_probs = vmap(vmap(loss, in_axes=(None, 0, 0)),
+                     in_axes=(None, 0, 0))(params, batch.info_state[agent_id],
+                                           batch.action[agent_id])
+    return -log_probs.sum(axis=-1).mean()
+
+  def update(train_state: TrainState,
+             batch: TransitionBatch) -> typing.Tuple[TrainState, typing.Dict]:
+    policy_params = train_state.policy_params[agent_id]
+    opt_state = train_state.policy_opt_states[agent_id]
+    loss = 0
+    for mini_batch in get_minibatches(batch, num_minibatches):
+      loss, policy_grads = jax.value_and_grad(loss_fn)(policy_params,
+                                                       mini_batch)
+      updates, opt_state = optimizer(policy_grads, opt_state)
+      policy_params = optax.apply_updates(train_state.policy_params[agent_id],
+                                          updates)
+
+    train_state = TrainState(
+      policy_params={
+        **train_state.policy_params, agent_id: policy_params
+      },
+      policy_opt_states={
+        **train_state.policy_opt_states, agent_id: opt_state
+      },
+      critic_params=deepcopy(train_state.critic_params),
+      critic_opt_states=deepcopy(train_state.critic_opt_states),
+    )
+    return train_state, {'loss': loss}
+
+  return update
+
+
+class OpponentShapingAgent(rl_agent.AbstractAgent):
```

Comment:
This is defining an OpenSpiel agent. Do you know if it runs in the general case, e.g. can two opponent shaping agents play tic-tac-toe in self-play?

I ask only because the examples seem to use the vectorized iterated matrix games, and I am curious if you tested on any sequential games.

Reply:
> Hi @axelbr, there will be a delay as I am now on holiday for two weeks. I will reply again after April 15th.

Ok, no problem. Have a nice holiday.



> This is defining an OpenSpiel agent. Do you know if it runs in the general case, e.g. can two opponent shaping agents play tic-tac-toe in self-play?

Yes. In fact, I use the regular repeated matrix games in the unit tests.

Reply:
I did not try it on other games than repeated matrix games though.

Reply:
It's ok. Just please document in the header of the opponent_shaping.py (and as a note inalgorithms.md) that it currently only works for simultaneous move games

Reply:
Ok, I added the info to both files.